There are a few principal functions reading data into R.
•
•
•
•
•
•
read.table , read.csv , for reading tabular data
readLines , for reading lines of a text file
source , for reading in R code files ( inverse of dump )
dget , for reading in R code files ( inverse of dput )
load , for reading in saved workspaces
unserialize , for reading single R objects in binary form

In [4]:
print (getwd)

function () 
.Internal(getwd())
<bytecode: 0x245a7e8>
<environment: namespace:base>


In [7]:
setwd("/home/nj/Path")

In [6]:
print (getwd)

function () 
.Internal(getwd())
<bytecode: 0x245a7e8>
<environment: namespace:base>


In [18]:
quora <-read.csv("test.csv")
str(quora)

'data.frame':	49999 obs. of  5 variables:
 $ id       : int  49999 50000 50001 50002 50003 50004 50005 50006 50007 50008 ...
 $ qid1     : int  88910 88912 88914 88916 88918 88919 88921 83290 88923 88925 ...
 $ qid2     : int  88911 88913 88915 88917 32373 88920 88922 58717 88924 88926 ...
 $ question1: Factor w/ 46000 levels " What high potential jobs are there other than computer science?",..: 42181 44442 30360 19033 38430 39106 19278 6153 22320 18897 ...
 $ question2: Factor w/ 45900 levels " Is the Haskell Programming From First Principles book worth buying?",..: 11739 44298 123 21142 38195 39070 19884 10566 23042 19469 ...


In [19]:
quora1 <-read.csv("test.csv",stringAsFactors=T)
str(quora)

ERROR: Error in read.table(file = file, header = header, sep = sep, quote = quote, : unused argument (stringAsFactors = TRUE)


In [40]:
write.csv(mtcars,"mt.csv",row.names=F)
read.csv("mt.csv")

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
1  21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
2  21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
3  22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
4  21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
5  18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
6  18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
7  14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
8  24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
9  22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
10 19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
11 17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
12 16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
13 17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
14 15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
15 10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
16 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
17 14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
18 32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
19 30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
20 33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
21 21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
22 15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
23 15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
24 13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
25 19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
26 27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
27 26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
28 30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
29 15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
30 19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
31 15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
32 21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

In [86]:
write.csv(mtcars,"mt.txt")
read.csv("mt.txt")

X                   mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
1  Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
2  Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
3  Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
4  Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
5  Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
6  Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
7  Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
8  Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
9  Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
10 Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
11 Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
12 Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
13 Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
14 Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
15 Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
16 Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
17 Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
18 Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
19 Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
20 Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
21 Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
22 Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
23 AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
24 Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
25 Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
26 Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
27 Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
28 Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
29 Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
30 Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
31 Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
32 Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

In [39]:
initial = read.table("mt.txt",nrows=10,skip=0)
initial

V1               
1                   
2  Mazda RX4        
3  Mazda RX4 Wag    
4  Datsun 710       
5  Hornet 4 Drive   
6  Hornet Sportabout
7  Valiant          
8  Duster 360       
9  Merc 240D        
10 Merc 230         
   V2                                                                 
1  ,"mpg","cyl","disp","hp","drat","wt","qsec","vs","am","gear","carb"
2  ,21,6,160,110,3.9,2.62,16.46,0,1,4,4                               
3  ,21,6,160,110,3.9,2.875,17.02,0,1,4,4                              
4  ,22.8,4,108,93,3.85,2.32,18.61,1,1,4,1                             
5  ,21.4,6,258,110,3.08,3.215,19.44,1,0,3,1                           
6  ,18.7,8,360,175,3.15,3.44,17.02,0,0,3,2                            
7  ,18.1,6,225,105,2.76,3.46,20.22,1,0,3,1                            
8  ,14.3,8,360,245,3.21,3.57,15.84,0,0,3,4                            
9  ,24.4,4,146.7,62,3.69,3.19,20,1,0,4,2                              
10 ,22.8,4,140.8,95,3.92,3.15,22.9,1,0,4,2

In [41]:
initial = read.table("mt.txt",nrows=10,skip=1)
initial

V1                V2                                      
1  Mazda RX4         ,21,6,160,110,3.9,2.62,16.46,0,1,4,4    
2  Mazda RX4 Wag     ,21,6,160,110,3.9,2.875,17.02,0,1,4,4   
3  Datsun 710        ,22.8,4,108,93,3.85,2.32,18.61,1,1,4,1  
4  Hornet 4 Drive    ,21.4,6,258,110,3.08,3.215,19.44,1,0,3,1
5  Hornet Sportabout ,18.7,8,360,175,3.15,3.44,17.02,0,0,3,2 
6  Valiant           ,18.1,6,225,105,2.76,3.46,20.22,1,0,3,1 
7  Duster 360        ,14.3,8,360,245,3.21,3.57,15.84,0,0,3,4 
8  Merc 240D         ,24.4,4,146.7,62,3.69,3.19,20,1,0,4,2   
9  Merc 230          ,22.8,4,140.8,95,3.92,3.15,22.9,1,0,4,2 
10 Merc 280          ,19.2,6,167.6,123,3.92,3.44,18.3,1,0,4,4

In [56]:
install.packages("xlsx")

also installing the dependencies ‘rJava’, ‘xlsxjars’

Warning message in install.packages("xlsx"):
“installation of package ‘rJava’ had non-zero exit status”Warning message in install.packages("xlsx"):
“installation of package ‘xlsxjars’ had non-zero exit status”Warning message in install.packages("xlsx"):
“installation of package ‘xlsx’ had non-zero exit status”Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [57]:
library(xlsx)

ERROR: Error in library(xlsx): there is no package called ‘xlsx’


In [77]:
input_excel = write.table(mtcars,"test_excel.xlsx")
read.xlsx("test_excel.xlsx")
\

ERROR: Error in eval(expr, envir, enclos): could not find function "read.xlsx"


For example, suppose I have a data frame with 1,500,000 rows and 120 columns, all of which are
numeric data. Roughly, how much memory is required to store this data frame? Well, on most
modern computers double precision floating point numbers are stored using 64 bits of memory, or
8 bytes. Given that information, you can do the following calculation:


Getting Data In and Out of R

#### 1,500,000 × 120 × 8 bytes/numeric

= 1,440,000,000 bytes
= 1,440,000,000 / 2 20 bytes/MB
= 1,373.29 MB
= 1.34 GB

In [64]:
install.packages("XML")
library(XML)
library(methods)


Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [66]:
xmldemo = xmlParseDoc(file="log.xml")


In [71]:
rootnode=xmlRoot(xmldemo)


In [76]:
print (rootnode)

<log4j:configuration xmlns:log4j="http://jakarta.apache.org/log4j/">
  <appender name="CONSOLE-WARN" class="org.apache.log4j.ConsoleAppender">
    <param name="target" value="System.err"/>
    <layout class="org.apache.log4j.PatternLayout">
      <param name="ConversionPattern" value="[%7r] %6p - %30.30c - %m \n"/>
    </layout>
    <filter class="org.apache.log4j.varia.LevelRangeFilter">
      <param name="LevelMin" value="WARN"/>
    </filter>
  </appender>

  <appender name="CONSOLE-DEBUG" class="org.apache.log4j.ConsoleAppender">
    <layout class="org.apache.log4j.PatternLayout">
      <param name="ConversionPattern" value="[%7r] %6p - %30.30c - %m \n"/>
    </layout>
    <filter class="org.apache.log4j.varia.LevelRangeFilter">
      <param name="LevelMin" value="DEBUG"/>
      <param name="LevelMax" value="DEBUG"/>
    </filter>
  </appender>

  <appender name="CONSOLE-ALL" class="org.apache.log4j.ConsoleAppender">
    <layout class="org.apache.log4j.PatternLayout">
      <param 

### Apply

In [79]:
x <-matrix(rnorm(30),nrow=5,ncol=6)
x

[,1]       [,2]       [,3]       [,4]       [,5]       [,6]       
[1,] -1.3214502  0.9757170  0.1659438 -0.3130428 -0.7793260  0.29719709
[2,] -0.1808931 -1.2765595 -0.7062806  0.7807515  0.2102781 -0.62780124
[3,] -2.0201855 -0.2800634 -0.3446207 -2.2597769  0.9777190  0.28669963
[4,] -0.1027343  0.3079296 -1.9467903  1.6858864 -0.6805446  0.04371526
[5,] -0.7381054  0.3362551  0.2926676  0.6004688 -0.8537934  0.40953476

In [81]:
y <-apply(x,2,sum)## 2 is for column, 1 is for row
y

[1] -4.36336857  0.06327878 -2.53908017  0.49428708 -1.12566681  0.40934551

In [82]:
y <-apply(x,1,sum)
y

[1] -0.9749611 -1.8005048 -3.6402279 -0.6925380  0.0470275

In [83]:
MyList <- list(1:5,20:25)
MyList

[[1]]
[1] 1 2 3 4 5

[[2]]
[1] 20 21 22 23 24 25

In [84]:
lapply(MyList,median)

[[1]]
[1] 3

[[2]]
[1] 22.5

In [87]:
sapply(MyList,median) # similar to lapply it tries to simplify the output

[1]  3.0 22.5

In [90]:
yx <- data.frame(a=1,b="a",c=1,2)
yx

a b c X2
1 1 a 1 2

In [96]:
a = dput(yx,"y.R")
a

a b c X2
1 1 a 1 2

In [93]:
ages <-c(25,67,34,25,23,66)
location <-c("urbal","metro","rural","urbal","metro","rural")
tapply(ages,location,mean)

metro rural urbal 
   45    50    25

In [98]:
x <- list (1,2,3)
serialize (x,NULL)

 [1] 58 0a 00 00 00 02 00 03 03 02 00 02 03 00 00 00 00 13 00 00 00 03 00 00 00
[26] 0e 00 00 00 01 3f f0 00 00 00 00 00 00 00 00 00 0e 00 00 00 01 40 00 00 00
[51] 00 00 00 00 00 00 00 0e 00 00 00 01 40 08 00 00 00 00 00 00

In [99]:
str(file)

function (description = "", open = "", blocking = TRUE, encoding = getOption("encoding"), 
    raw = FALSE, method = getOption("url.method", "default"))  


In [102]:
con <-file("log.txt")


In [103]:
open(con,"r")
data <-read.csv(con)

Warning message in open.connection(con, "r"):
“cannot open file 'log.txt': No such file or directory”

ERROR: Error in open.connection(con, "r"): cannot open the connection
